url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/Eval2.java

```diff
@@ -0,0 +1,1640 @@
+implements Eval2.IFrameHandler
+{
+    public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        // No stack frame space needed for atoms.
+    }
+
+    public void handleFrame(IEval aEval, Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        final Object lExpr = aFrame.getExpr();
+        final IContext lCtx = aFrame.getCtx();
+
+        if(lExpr instanceof String && ((String)lExpr).startsWith("$"))
+        {
+            // Syntactic sugar. $name is equivalent to (get name).
+            aFrame.setResult(lCtx.getBinding(((String)lExpr).substring(1)));
+        }
+        else
+        {
+            // All other values except lists evaluate to
+            // themselves! This is quite different from the original LISP.
+            // Our goal is to have a handy and flexible REPL that can handle model objects or "handles"
+            // that do not have a list representation.
+            aFrame.setResult(lExpr);
+        }
+    }
+}
+
+class PairHandler
+implements Eval2.IFrameHandler
+{
+    public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        // Allocate stack data.
+        aFrame.allocateData(2);
+    }
+
+    public void handleFrame(IEval aEval, Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        final Object[] lData = aFrame.getData();
+        final IContext lCtx = aFrame.getCtx();
+        final Pair lPair = (Pair) aFrame.getExpr();
+        switch(aFrame.getDataptr())
+        {
+        case 0:
+            // Put the left side on the stack for evaluation.
+            aStack.push(lPair.getLeft(), lCtx);
+            return;
+        case 1:
+            // Put the right side on the stack for evaluation.
+            aStack.push(lPair.getRight(), lCtx);
+            return;
+        default:
+            // Create a new pair and add it as the result of the current
+            // expression to the stack frame.
+            aFrame.setResult(new Pair(lData[0], lData[1]));
+        }
+    }
+}
+
+class StandardHandler
+implements Eval2.IFrameHandler
+{
+    private CommandRepository commands = new CommandRepository();
+
+    public StandardHandler()
+    {
+    }
+
+    public CommandRepository getCommands()
+    {
+        return commands;
+    }
+
+    public void setCommands(CommandRepository aCommands)
+    {
+        commands = aCommands;
+    }
+
+    public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        final List lLstExpr = (List) aFrame.getExpr();
+        final int lLstSize = lLstExpr.size();
+        // Allocate enough space to evaluate all elements of the expression.
+        // So the space should be as large as the original expression.
+        aFrame.allocateData(lLstSize);
+    }
+
+    public void handleFrame(IEval aEval, Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 31. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.